Adding OpenSSL dependency to Boost::boost

tatraian opened the pull request:
websocketpp required OpenSSL for Boost::asio (that is a header only library). Now Boost::boost can be linked together with OpenSSL libraries depending on BOOST_CONFIG_LINK_OPENSSL variable.

ruslo commented on cmake/templates/BoostConfig.cmake.in:
```diff
@@ -39,6 +39,18 @@ if(NOT TARGET "Boost::boost")
   endif()
+
+  if(BOOST_CONFIG_LINK_OPENSSL)
+    hunter_add_package(OpenSSL)
```
This file used **after** Boost already built. So I guess the fact that OpenSSL is linked to Boost not used on build stage of Boost? If not then user should call `hunter_add_package(OpenSSL)` himself.

tatraian replied:
You are right I put it somewhere else, but I didn't find where. I enabled OpenSSL to Boost::boost "package" only. This is the header only part of boost. So I think install after boost is not elegant but works for me. I used the patched version of websocketpp (https://github.com/tatraian/websocketpp/tree/0.7.0-hunter).

ruslo replied:
If `websocketpp` use boost.asio with OpenSSL support (and this fact doesn't affect how Boost build) then it can be resolved by:

```cmake
hunter_add_package(Boost)
hunter_add_package(OpenSSL)

# find_*

target_link_libraries(foo PUBLIC Boost::boost ${OpenSSL_LIBRARIES})
```

ruslo replied:
Also see note about OpenSSL imported targets: https://github.com/hunter-packages/websocketpp/pull/1#issuecomment-213371486

This one is important and should be solved first.

tatraian commented:
I removed OpenSSL finding from BoostConfig. I moved the latest FindOpenSSL from cmake repository. I had to modify it a bit, since 
`SelectLibraryConfigurations` and `FindPackageHandleStandardArgs` were included as local source file. But no other modification.
If I use this library and websocketpp add OpenSSL to boost than I get link error for `dl` library. Do you have idea how can I fix that?

ruslo commented on cmake/find/FindOpenSSL.cmake:
```diff
@@ -25,15 +54,57 @@
 #  License text for the above reference.)
 
-if(HUNTER_STATUS_DEBUG)
-  message("[hunter] Custom FindOpenSSL module")
```
Please keep this one.

ruslo commented on cmake/find/FindOpenSSL.cmake:
```diff
@@ -25,15 +54,57 @@
 find_path(OPENSSL_INCLUDE_DIR
   NAMES
     openssl/ssl.h
+  ${_OPENSSL_ROOT_HINTS_AND_PATHS}
```
All hints should be removed. Only `OPENSSL_ROOT` which is set by Hunter must be used.

ruslo commented:
> If I use this library and websocketpp adds OpenSSL to boost than I get link error for dl library. Do you have idea how can I fix that?

What is the source of undefined reference OpenSSL? Asio? Websocketpp? You can use [CMAKE_DL_LIBS](https://cmake.org/cmake/help/latest/variable/CMAKE_DL_LIBS.html).

> I moved the latest FindOpenSSL from cmake repository

It would be nice to test new functionality, please add `examples/OpenSSL-imported` example similar to `examples/OpenSSL`: https://github.com/ruslo/hunter/blob/master/examples/OpenSSL/CMakeLists.txt

tatraian commented:
Example has been added and other changes has been fixed.

I created a wersocketpp example in hunter with openssl. That was the project which didn't find DL libraries. If I add CMAKE_DL_LIBS to websocketpp that will works on Windows as well?

ruslo commented on cmake/find/FindOpenSSL.cmake:
```diff
@@ -1,19 +1,48 @@
 # Copyright 2006-2009 Kitware, Inc.
 # Copyright 2006 Alexander Neundorf <neundorf@kde.org>
 # Copyright 2009-2011 Mathieu Malaterre <mathieu.malaterre@gmail.com>
-# Copyright 2015 Alexander Lamaison <alexander.lamaison@gmail.com>
```
Please keep this one

ruslo commented on cmake/find/FindOpenSSL.cmake:
```diff
@@ -54,53 +128,63 @@ if(WIN32 AND NOT CYGWIN)
         libeay32d
-      HINTS
-        "${OPENSSL_ROOT}"
+      ${_OPENSSL_ROOT_HINTS_AND_PATHS}
```
`OPENSSL_ROOT` should be used here too. Effectively this patch need to be applied:
* https://github.com/ruslo/hunter/commit/2375c2cc1d0ccbf7e5590fd9ab3053f5c9c52731

ruslo commented:
> If I add CMAKE_DL_LIBS to websocketpp that will works on Windows as well?

I think yes. This variable probably will be empty. Will be checked by Windows testing anyway.

tatraian commented:
I merged @alamaison patch into FindOpenSSL.cmake. Both websocketpp and OpenSSL-imported example run at me.

ruslo commented:
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.401
* https://travis-ci.org/ingenue/hunter/builds/125786820
